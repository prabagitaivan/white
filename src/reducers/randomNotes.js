import { createActions, handleActions } from 'redux-actions'

const defaultState = {
  data: [],
  requesting: true,
  error: null
}

export const {
  randomNotes: { request, success, failure }
} = createActions({
  RANDOM_NOTES: {
    REQUEST: action => action,
    SUCCESS: action => action,
    FAILURE: action => action
  }
})

export default handleActions(
  {
    [request]: state => ({ ...state, requesting: true, error: null }),
    [success]: (state, { payload: data = [] }) => ({
      ...state,
      data,
      requesting: false,
      error: null
    }),
    [failure]: (state, { payload: error = 'error' }) => ({
      ...state,
      requesting: false,
      error
    })
  },
  defaultState
)
