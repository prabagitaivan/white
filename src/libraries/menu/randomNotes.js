import { FlashOn, Loop } from '@material-ui/icons'
import { shuffle, random } from 'lodash'
import store from '../../stores'
import { edit } from '../../reducers/randomNotes'

export default [
  {
    Icon: Loop,
    text: 'Shuffle Notes',
    action: () => {
      const { data } = store.getState().randomNotes
      if (data && data.length > 0) {
        store.dispatch(edit(shuffle(data)))
      }
    }
  },
  {
    Icon: FlashOn,
    text: 'Open Random',
    action: () => {
      const { data } = store.getState().randomNotes
      if (data && data.length > 0) {
        window.open(data[random(data.length - 1)].url)
      }
    }
  }
]
