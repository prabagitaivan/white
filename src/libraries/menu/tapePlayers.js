import { Loop } from '@mui/icons-material'
import { shuffle } from 'lodash'
import store from '../../stores'
import { edit } from '../../reducers/tapePlayers'

const options = [
  {
    Icon: Loop,
    text: 'shuffle-players',
    action: () => {
      const { data } = store.getState().tapePlayers
      if (data && data.length > 0) {
        store.dispatch(edit(shuffle(data)))
      }
    }
  }
]

export default options
