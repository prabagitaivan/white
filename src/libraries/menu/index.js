import randomNotes from './randomNotes'

const menu = {
  Home: {
    name: 'Random Notes',
    options: randomNotes
  }
}

export default menu
