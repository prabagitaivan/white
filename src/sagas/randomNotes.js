import { takeLatest, call, put } from 'redux-saga/effects'

import { request, success, failure } from '../reducers/randomNotes'
import { getRandomNotes } from '../libraries/firebase/database'

export function * requestSaga () {
  try {
    const data = yield call(getRandomNotes)
    yield put(success(data))
  } catch (error) {
    yield put(failure(error.message))
  }
}

export function * watchRequest () {
  yield takeLatest(request.toString(), requestSaga)
}

export default [watchRequest]
