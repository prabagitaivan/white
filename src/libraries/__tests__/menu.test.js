import _ from 'lodash'
import store from '../../stores'
import menu from '../menu'
import randomNotes from '../menu/randomNotes'

function mockRandomNotesStore ({ data = [] }) {
  jest.spyOn(store, 'getState').mockReturnValue({ randomNotes: { data } })
}

describe('libraries menu', () => {
  describe('data', () => {
    it('return all correct menu', () => {
      expect(menu.Home).toEqual({ name: 'Random Notes', options: randomNotes })
    })
  })
  describe('random notes', () => {
    beforeEach(() => {
      jest.spyOn(_, 'shuffle').mockReturnValue([])
      jest.spyOn(_, 'random').mockReturnValue(0)
      jest.spyOn(store, 'dispatch').mockReturnValue()
      jest.spyOn(window, 'open').mockReturnValue()
    })
    afterEach(() => {
      _.shuffle.mockRestore()
      _.random.mockRestore()
      store.dispatch.mockRestore()
      store.getState.mockRestore()
      window.open.mockRestore()
    })

    it('return all correct icon and order', () => {
      mockRandomNotesStore({})
      const icons = randomNotes.map(note => note.Icon)
      expect(icons).toMatchSnapshot()
    })
    it('return all correct text and order', () => {
      mockRandomNotesStore({})
      const texts = randomNotes.map(note => note.text)
      expect(texts[0]).toEqual('Shuffle Notes')
      expect(texts[1]).toEqual('Open Random')
    })
    it('shuffle notes if data exists', () => {
      mockRandomNotesStore({ data: [1, 2] })
      const shuffleOption = randomNotes[0]
      shuffleOption.action()
      expect(_.shuffle).toHaveBeenCalledTimes(1)
      expect(store.dispatch).toHaveBeenCalledTimes(1)
    })
    it('not shuffle notes if data empty', () => {
      mockRandomNotesStore({})
      const shuffleOption = randomNotes[0]
      shuffleOption.action()
      expect(_.shuffle).not.toHaveBeenCalled()
      expect(store.dispatch).not.toHaveBeenCalled()
    })
    it('open 1 random note if data exists', () => {
      mockRandomNotesStore({ data: [{ url: 1 }, { url: 2 }] })
      const randomOption = randomNotes[1]
      randomOption.action()
      expect(_.random).toHaveBeenCalledTimes(1)
      expect(window.open).toHaveBeenCalledTimes(1)
    })
    it('not open 1 random note if data empty', () => {
      mockRandomNotesStore({})
      const randomOption = randomNotes[1]
      randomOption.action()
      expect(_.random).not.toHaveBeenCalled()
      expect(window.open).not.toHaveBeenCalled()
    })
  })
})
