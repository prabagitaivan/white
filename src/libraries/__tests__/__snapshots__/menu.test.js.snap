// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`libraries menu random notes return all correct icon and order 1`] = `
Array [
  Object {
    "$$typeof": Symbol(react.memo),
    "compare": null,
    "displayName": "LoopIcon",
    "muiName": "SvgIcon",
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
  },
  Object {
    "$$typeof": Symbol(react.memo),
    "compare": null,
    "displayName": "FlashOnIcon",
    "muiName": "SvgIcon",
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
  },
]
`;
